File: src/main/java/ru/isb/bot/clients/NextcloudClientImpl.java
package ru.isb.bot.clients;

import lombok.SneakyThrows;
import lombok.extern.log4j.Log4j2;
import org.apache.http.StatusLine;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPut;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.FileEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import ru.isb.bot.utils.StringUtils;

import java.io.File;
import java.io.IOException;
import java.util.Base64;

@Log4j2
@Component
public class NextcloudClientImpl implements NextcloudClient {

    @Value("${nextcloud.host}")
    private String NEXTCLOUD_HOST;

    @Value("${nextcloud.user.name}")
    private String NEXTCLOUD_USER_NAME;

    @Value("${nextcloud.user.pass}")
    private String NEXTCLOUD_USER_PASS;

    @Value("${nextcloud.path.to.file}")
    private String NEXTCLOUD_PATH_TO_FILE;
    private static final String NEXTCLOUD_URL_PATH = "/remote.php/dav/files/isb/";

    @SneakyThrows
    public void uploadFile(File file, String fileName) {
        fileName = StringUtils.replaceSpace(fileName);

        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
            HttpPut request = new HttpPut(NEXTCLOUD_HOST + NEXTCLOUD_URL_PATH + NEXTCLOUD_PATH_TO_FILE + "/" + fileName);
            String encodedCredentials = createCredentials(request);
            FileEntity fileEntity = createEntityFile(request, file);

            try (CloseableHttpResponse response = httpClient.execute(request)) {
                StatusLine statusLine = response.getStatusLine();
                int statusCode = statusLine.getStatusCode();
                log.info("File upload status code: " + statusCode);

                if (statusCode == 204) {
                    uploadFile(file, StringUtils.copyFileName(fileName));
                } else if (statusCode == 302) {
                    redirect(response, fileEntity, encodedCredentials, httpClient);
                }
            }
        } catch (IOException ex) {
            log.error(ex.getMessage(), ex);
        }
    }

    @SneakyThrows
    private void redirect(CloseableHttpResponse response, FileEntity fileEntity, String encodedCredentials, CloseableHttpClient httpClient) {
        String redirectUrl = response.getFirstHeader("Location").getValue();
        log.info("Redirect URL: " + redirectUrl);

        HttpPut redirectRequest = new HttpPut(redirectUrl);
        redirectRequest.addHeader("Authorization", "Basic " + encodedCredentials);
        redirectRequest.setEntity(fileEntity);

        try (CloseableHttpResponse redirectResponse = httpClient.execute(redirectRequest)) {
            StatusLine redirectStatusLine = redirectResponse.getStatusLine();
            int redirectStatusCode = redirectStatusLine.getStatusCode();
            log.info("Redirect status code: " + redirectStatusCode);
        }
    }

    private String createCredentials(HttpPut request) {
        String credentials = NEXTCLOUD_USER_NAME + ":" + NEXTCLOUD_USER_PASS;
        String encodedCredentials = Base64.getEncoder().encodeToString(credentials.getBytes());
        request.addHeader("Authorization", "Basic " + encodedCredentials);
        return encodedCredentials;
    }

    private FileEntity createEntityFile(HttpPut request, File file) {
        FileEntity fileEntity = new FileEntity(file, ContentType.DEFAULT_BINARY);
        request.setEntity(fileEntity);
        return fileEntity;
    }
}

File: src/main/java/ru/isb/bot/clients/NextcloudClientImpl.kt
package ru.isb.bot.clients

import lombok.SneakyThrows
import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.client.methods.HttpPut
import org.apache.http.entity.ContentType
import org.apache.http.entity.FileEntity
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClients
import org.apache.logging.log4j.kotlin.Logging
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import ru.isb.bot.utils.PathFileUtils.Companion.copyFileName
import ru.isb.bot.utils.PathFileUtils.Companion.replaceSpace
import java.io.File
import java.io.IOException
import java.util.*

@Component
class NextcloudClientImpl : NextcloudClient, Logging {

    @Value("\${nextcloud.host}")
    private val NEXTCLOUD_HOST: String = ""

    @Value("\${nextcloud.user.name}")
    private val NEXTCLOUD_USER_NAME: String = ""

    @Value("\${nextcloud.user.pass}")
    private val NEXTCLOUD_USER_PASS: String = ""

    @Value("\${nextcloud.path.to.file}")
    private val NEXTCLOUD_PATH_TO_FILE: String = ""
    private val NEXTCLOUD_URL_PATH = "/remote.php/dav/files/isb/"

    @SneakyThrows
    override fun uploadFile(file: File, fileName: String) {
        val replacedFileName = replaceSpace(fileName)

        try {
            HttpClients.createDefault().use { httpClient ->
                val request =
                    HttpPut("$NEXTCLOUD_HOST$NEXTCLOUD_URL_PATH$NEXTCLOUD_PATH_TO_FILE/$replacedFileName")
                val encodedCredentials = createCredentials(request)
                val fileEntity = createEntityFile(request, file)
                httpClient.execute(request).use { response ->
                    val statusLine = response.statusLine
                    val statusCode = statusLine.statusCode
                    logger.info("File upload status code: $statusCode")
                    if (statusCode == 204) {
                        uploadFile(file, copyFileName(replacedFileName))
                    } else if (statusCode == 302) {
                        redirect(response, fileEntity, encodedCredentials, httpClient)
                    }
                }
            }
        } catch (ex: IOException) {
            logger.error(ex.message ?: "Ошибка запроса к облаку", ex)
        }
    }

    @SneakyThrows
    private fun redirect(
        response: CloseableHttpResponse,
        fileEntity: FileEntity,
        encodedCredentials: String,
        httpClient: CloseableHttpClient
    ) {
        val redirectUrl = response.getFirstHeader("Location").value

        logger.info("Redirect URL: $redirectUrl")

        val redirectRequest = HttpPut(redirectUrl)
        redirectRequest.addHeader("Authorization", "Basic $encodedCredentials")
        redirectRequest.entity = fileEntity
        httpClient.execute(redirectRequest).use { redirectResponse ->
            logger.info("Redirect status code: ${redirectResponse.statusLine.statusCode}")
        }
    }

    private fun createCredentials(request: HttpPut): String {
        val credentials = "$NEXTCLOUD_USER_NAME:$NEXTCLOUD_USER_PASS"
        val encodedCredentials = Base64.getEncoder().encodeToString(credentials.toByteArray())
        request.addHeader("Authorization", "Basic $encodedCredentials")
        return encodedCredentials
    }

    private fun createEntityFile(request: HttpPut, file: File): FileEntity {
        val fileEntity = FileEntity(file, ContentType.DEFAULT_BINARY)
        request.entity = fileEntity
        return fileEntity
    }

}